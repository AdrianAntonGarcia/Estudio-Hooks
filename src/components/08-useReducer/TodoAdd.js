import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = React.memo(({ handleAddTodo }) => {
	console.log('Renderizado TodoAdd');
	const [{ description }, handleInputChange, reset] = useForm({
		description: "",
	});

	const handleSubmit = (e) => {
		e.preventDefault();
		if (description.trim().length <= 1) {
			return;
		}
		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false,
		};
		handleAddTodo(newTodo);
		reset();
	};

	return (
		<>
			<h4>Agregar TODO</h4>
			<hr />
			<form
				onSubmit={(e) => {
					handleSubmit(e, reset);
				}}
			>
				<input
					type="text"
					name="description"
					className="form-control"
					placeholder="Aprender ..."
					autoComplete="off"
					value={description}
					onChange={handleInputChange}
				/>
				<button
					type="submit"
					className="btn btn-outline-primary mt-1 btn-block"
				>
					Agregar
				</button>
			</form>
		</>
	);
});
