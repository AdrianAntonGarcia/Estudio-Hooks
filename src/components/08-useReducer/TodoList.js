import React from "react";
import { TodoListItem } from "./TodoListItem";

export const TodoList = React.memo(({ todos, handleDelete, handleToggle }) => {
	console.log('Renderizado TodoList');
	return (
		<>
			{/**
			 * TodoList, todos, handleDelete, handleToggle
			 */}
			<ul className="list-group list-group-flush">
				{todos.map((todo, i) => (
					<TodoListItem
						key={todo.id}
						todo={todo}
						index={i}
						handleDelete={handleDelete}
						handleToggle={handleToggle}
					/>
				))}
			</ul>
		</>
	);
});
