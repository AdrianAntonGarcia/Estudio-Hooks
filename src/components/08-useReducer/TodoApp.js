import React, { useReducer, useEffect, useCallback } from "react";
import { todoReducer } from "./todoReducer";
import "./styles.css";
// import { useForm } from "../../hooks/useForm";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

/**
 * Lo que retorne la función init es lo que será el initial state,
 * se puede usar solo con dos argumentos pasando el initial state
 * como segundo.
 */
const init = () => {
	// Si retorna null devuelve el arreglo vacio
	return JSON.parse(localStorage.getItem("todos")) || [];
	// return [
	// 	{
	// 		id: new Date().getTime(),
	// 		desc: "Aprender react",
	// 		done: false,
	// 	},
	// ];
};

export const TodoApp = () => {
	const [todos, dispatch] = useReducer(todoReducer, [], init);

	/**
	 * Cuándo los todos cambien, grabamos en el local storage
	 */
	useEffect(() => {
		localStorage.setItem("todos", JSON.stringify(todos));
	}, [todos]);

	const handleDelete = useCallback(
		(todoId) => {
			const action = {
				type: "delete",
				payload: todoId,
			};
			dispatch(action);
		},
		[]
	);
	// const handleDelete = (todoId) => {
	// 	const action = {
	// 		type: "delete",
	// 		payload: todoId,
	// 	};
	// 	dispatch(action);
	// };

	const handleToggle = useCallback(
		(todoId) => {
			dispatch({ type: "toggle", payload: todoId });
		},
		[]
	);

	// const handleToggle = (todoId) => {
	// 	dispatch({ type: "toggle", payload: todoId });
	// };

	const handleAddTodo = useCallback(
		(newTodo) => {
			dispatch({
				type: "add",
				payload: newTodo,
			});
		},
		[]
	);
	// const handleAddTodo = (newTodo) => {
	// 	dispatch({
	// 		type: "add",
	// 		payload: newTodo,
	// 	});
	// };

	return (
		<div>
			<h1>TodoApp ({todos.length})</h1>
			<hr />
			<div className="row">
				<div className="col-7">
					<TodoList
						todos={todos}
						handleDelete={handleDelete}
						handleToggle={handleToggle}
					/>
				</div>
				<div className="col-5">
					<TodoAdd handleAddTodo={handleAddTodo} />
				</div>
			</div>
		</div>
	);
};

// const handleSubmit = (e) => {
// 	if (description.trim().length <= 1) {
// 		return;
// 	}
// 	e.preventDefault();
// 	const newTodo = {
// 		id: new Date().getTime(),
// 		desc: description,
// 		done: false,
// 	};

// 	const action = {
// 		type: "add",
// 		payload: newTodo,
// 	};

// 	dispatch(action);
// 	reset();
// };
