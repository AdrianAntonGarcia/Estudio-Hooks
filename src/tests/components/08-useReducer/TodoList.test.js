import React from "react";
import { shallow } from "enzyme";
import { TodoList } from "../../../components/08-useReducer/TodoList";
import { demoTodos } from "../../fixtures/demoTodos";

describe("Pruebas en TodoList", () => {
	const handleDelete = jest.fn();
	const handleToggle = jest.fn();
	const wrapper = shallow(
		<TodoList
			todos={demoTodos}
			handleDelete={handleDelete}
			handleToggle={handleToggle}
		/>
	);
	test("Debe de mostrarse correctamente", () => {
		expect(wrapper).toMatchSnapshot();
	});
	test("Debe de tener dos <TodoListItem/>", () => {
		console.log(wrapper.text());

		expect(wrapper.find("Memo()").length).toBe(demoTodos.length);
		expect(wrapper.find("Memo()").at(0).prop("handleDelete"))
			.toEqual(expect.any(Function));
	});
});
