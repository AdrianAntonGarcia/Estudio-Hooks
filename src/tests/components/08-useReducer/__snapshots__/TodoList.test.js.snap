// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoList Debe de mostrarse correctamente 1`] = `
<Fragment>
  <ul
    className="list-group list-group-flush"
  >
    <Memo()
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      index={0}
      key="1"
      todo={
        Object {
          "desc": "Aprender React",
          "done": false,
          "id": 1,
        }
      }
    />
    <Memo()
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      index={1}
      key="2"
      todo={
        Object {
          "desc": "Aprender Mongo",
          "done": false,
          "id": 2,
        }
      }
    />
  </ul>
</Fragment>
`;
